File: examples-js/basic_examples.js
/**
 * Basic Examples - Copilot Simulator
 * 
 * This file demonstrates basic usage of the Copilot simulator
 * with simple, common coding scenarios.
 */

const { CopilotSimulator } = require('../copilot_simulator');

function exampleFunctionDocstring() {
    console.log('\n' + '='.repeat(70));
    console.log('EXAMPLE: Function Documentation Suggestion');
    console.log('='.repeat(70));

    const code = `function calculateAverage(numbers) {`;

    const simulator = new CopilotSimulator();
    const suggestions = simulator.getSuggestions(
        code,
        [1, 0],
        'math_utils.js',
        false
    );

    console.log(`\nCode context:\n${code}`);
    console.log(`\nCursor position: Line 1, Column 0`);
    console.log(`\nTop suggestions:`);
    suggestions.slice(0, 3).forEach((sug, i) => {
        console.log(`\n${i + 1}. Confidence: ${(sug.confidence * 100).toFixed(0)}%`);
        console.log(`   Reasoning: ${sug.reasoning}`);
        console.log(`   Suggestion:\n${sug.text}`);
    });
}

function exampleClassInitialization() {
    console.log('\n' + '='.repeat(70));
    console.log('EXAMPLE: Class Initialization Suggestion');
    console.log('='.repeat(70));

    const code = `class UserAccount {`;

    const simulator = new CopilotSimulator();
    const suggestions = simulator.getSuggestions(
        code,
        [1, 0],
        'models.js',
        false
    );

    console.log(`\nCode context:\n${code}`);
    console.log(`\nCursor position: Line 1, Column 0`);
    console.log(`\nTop suggestions:`);
    suggestions.slice(0, 3).forEach((sug, i) => {
        console.log(`\n${i + 1}. Confidence: ${(sug.confidence * 100).toFixed(0)}%`);
        console.log(`   Reasoning: ${sug.reasoning}`);
        console.log(`   Suggestion:\n${sug.text}`);
    });
}

function exampleCommentToCode() {
    console.log('\n' + '='.repeat(70));
    console.log('EXAMPLE: Comment to Code Suggestion');
    console.log('='.repeat(70));

    const code = `// Loop through all users and validate their email addresses`;

    const simulator = new CopilotSimulator();
    const suggestions = simulator.getSuggestions(
        code,
        [1, 0],
        'validation.js',
        false
    );

    console.log(`\nCode context:\n${code}`);
    console.log(`\nCursor position: Line 1, Column 0`);
    console.log(`\nTop suggestions:`);
    suggestions.slice(0, 3).forEach((sug, i) => {
        console.log(`\n${i + 1}. Confidence: ${(sug.confidence * 100).toFixed(0)}%`);
        console.log(`   Reasoning: ${sug.reasoning}`);
        console.log(`   Suggestion:\n${sug.text}`);
    });
}

function exampleWithContext() {
    console.log('\n' + '='.repeat(70));
    console.log('EXAMPLE: Suggestions with Rich Context');
    console.log('='.repeat(70));

    const code = `const fs = require('fs');
const path = require('path');

class DataProcessor {
    constructor(configPath) {
        this.configPath = configPath;
        this.data = [];
    }

    loadData(filePath) {
        // Load JSON data from file
`;

    const simulator = new CopilotSimulator();
    const suggestions = simulator.getSuggestions(
        code,
        [11, 0],
        'processor.js',
        false
    );

    const lines = code.split('\n');
    console.log(`\nCode context (showing last 5 lines):`);
    console.log(lines.slice(-5).join('\n'));
    console.log(`\nCursor position: Line 11, Column 0`);
    console.log(`\nTop suggestions:`);
    suggestions.slice(0, 3).forEach((sug, i) => {
        console.log(`\n${i + 1}. Confidence: ${(sug.confidence * 100).toFixed(0)}%`);
        console.log(`   Reasoning: ${sug.reasoning}`);
        console.log(`   Suggestion:\n${sug.text}`);
    });
}

function main() {
    console.log('\n' + '='.repeat(70));
    console.log('COPILOT SIMULATOR - BASIC EXAMPLES');
    console.log('='.repeat(70));
    console.log('\nThese examples demonstrate how Copilot generates suggestions');
    console.log('in common coding scenarios.\n');

    exampleFunctionDocstring();
    exampleClassInitialization();
    exampleCommentToCode();
    exampleWithContext();

    console.log('\n' + '='.repeat(70));
    console.log('ALL EXAMPLES COMPLETE');
    console.log('='.repeat(70));
    console.log('\nKey Takeaways:');
    console.log('1. Copilot analyzes the code context around your cursor');
    console.log('2. It recognizes common patterns (functions, classes, comments)');
    console.log('3. It generates multiple suggestions with confidence scores');
    console.log('4. Higher confidence suggestions are ranked first');
    console.log('5. The suggestions are based on learned patterns from training data');
    console.log('\n');
}

if (require.main === module) {
    main();
}

module.exports = { main };

File: examples/basic_examples.py
"""
Basic Examples - Copilot Simulator

This file demonstrates basic usage of the Copilot simulator
with simple, common coding scenarios.
"""

import sys
import os

# Add parent directory to path to import copilot_simulator
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from copilot_simulator import CopilotSimulator


def example_function_docstring():
    """Example: Suggesting docstring after function definition"""
    print("\n" + "=" * 70)
    print("EXAMPLE: Function Docstring Suggestion")
    print("=" * 70)
    
    code = """def calculate_average(numbers):"""
    
    simulator = CopilotSimulator()
    suggestions = simulator.get_suggestions(
        file_content=code,
        cursor_position=(1, 0),
        file_path="math_utils.py",
        verbose=False
    )
    
    print(f"\nCode context:\n{code}")
    print(f"\nCursor position: Line 1, Column 0")
    print(f"\nTop suggestions:")
    for i, sug in enumerate(suggestions[:3], 1):
        print(f"\n{i}. Confidence: {sug.confidence:.2%}")
        print(f"   Reasoning: {sug.reasoning}")
        print(f"   Suggestion:\n{sug.text}")


def example_class_initialization():
    """Example: Suggesting __init__ after class definition"""
    print("\n" + "=" * 70)
    print("EXAMPLE: Class Initialization Suggestion")
    print("=" * 70)
    
    code = """class UserAccount:"""
    
    simulator = CopilotSimulator()
    suggestions = simulator.get_suggestions(
        file_content=code,
        cursor_position=(1, 0),
        file_path="models.py",
        verbose=False
    )
    
    print(f"\nCode context:\n{code}")
    print(f"\nCursor position: Line 1, Column 0")
    print(f"\nTop suggestions:")
    for i, sug in enumerate(suggestions[:3], 1):
        print(f"\n{i}. Confidence: {sug.confidence:.2%}")
        print(f"   Reasoning: {sug.reasoning}")
        print(f"   Suggestion:\n{sug.text}")


def example_comment_to_code():
    """Example: Suggesting code based on comment"""
    print("\n" + "=" * 70)
    print("EXAMPLE: Comment to Code Suggestion")
    print("=" * 70)
    
    code = """# Loop through all users and validate their email addresses"""
    
    simulator = CopilotSimulator()
    suggestions = simulator.get_suggestions(
        file_content=code,
        cursor_position=(1, 0),
        file_path="validation.py",
        verbose=False
    )
    
    print(f"\nCode context:\n{code}")
    print(f"\nCursor position: Line 1, Column 0")
    print(f"\nTop suggestions:")
    for i, sug in enumerate(suggestions[:3], 1):
        print(f"\n{i}. Confidence: {sug.confidence:.2%}")
        print(f"   Reasoning: {sug.reasoning}")
        print(f"   Suggestion:\n{sug.text}")


def example_with_context():
    """Example: Suggestions with richer context"""
    print("\n" + "=" * 70)
    print("EXAMPLE: Suggestions with Rich Context")
    print("=" * 70)
    
    code = """import json
import os
from typing import List, Dict

class DataProcessor:
    def __init__(self, config_path: str):
        self.config_path = config_path
        self.data = []
    
    def load_data(self, file_path: str):
        # Load JSON data from file
"""
    
    simulator = CopilotSimulator()
    suggestions = simulator.get_suggestions(
        file_content=code,
        cursor_position=(11, 0),
        file_path="processor.py",
        verbose=False
    )
    
    print(f"\nCode context (showing last 5 lines):")
    print('\n'.join(code.split('\n')[-5:]))
    print(f"\nCursor position: Line 11, Column 0")
    print(f"\nTop suggestions:")
    for i, sug in enumerate(suggestions[:3], 1):
        print(f"\n{i}. Confidence: {sug.confidence:.2%}")
        print(f"   Reasoning: {sug.reasoning}")
        print(f"   Suggestion:\n{sug.text}")


def main():
    """Run all basic examples"""
    print("\n" + "=" * 70)
    print("COPILOT SIMULATOR - BASIC EXAMPLES")
    print("=" * 70)
    print("\nThese examples demonstrate how Copilot generates suggestions")
    print("in common coding scenarios.\n")
    
    example_function_docstring()
    example_class_initialization()
    example_comment_to_code()
    example_with_context()
    
    print("\n" + "=" * 70)
    print("ALL EXAMPLES COMPLETE")
    print("=" * 70)
    print("\nKey Takeaways:")
    print("1. Copilot analyzes the code context around your cursor")
    print("2. It recognizes common patterns (functions, classes, comments)")
    print("3. It generates multiple suggestions with confidence scores")
    print("4. Higher confidence suggestions are ranked first")
    print("5. The suggestions are based on learned patterns from training data")
    print("\n")


if __name__ == "__main__":
    main()
